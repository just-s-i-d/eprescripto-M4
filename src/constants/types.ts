export type InfoCardDetailsType =
  | {
      totalCount: number;
      title: string;
    }[]
  | undefined;

export type ErrorType = {
  message: string;
  type: string;
};
export type AppointmentDataType = {
  pId: string;
  email: string;
  lastVisited: string;
  avatar: string;
  gender: string;
  key: number;
  pName: string;
  timeSlot: string;
  contact: string;
  status: string;
  profilePic: string;
  referrer: string;
  date: string;
};

export type AppointmentsDataType = AppointmentDataType[] | undefined;

export type PrescriptionDataType = {
  key: React.Key;
  prescriptionId: string;
  name: string;
  duration: number;
  date: string;
  status: "Active" | "Completed" | "On Hold";
  notes: string;
};

export type PrescrtionsDataType = PrescriptionDataType[] | undefined;

export type ChartDataType = {
  labels: string[];
  data: number[];
  totalCount: number;
};

export type LineChartDataType = {
  labels: string[];
  data: number[];
};

export type LineChartApiResDataType = {
  seven: number[];
  ten: number[];
  thirty: number[];
};

export type ChartApiDataType = {
  pieChartData: ChartDataType;
  barChartData: ChartDataType;
  lineChartData: LineChartApiResDataType;
};

export type DoctorAppointmentsDataType = AppointmentDataType[];

export type ReviewDataType = {
  pId: string;
  pName: string;
  email: string;
  rating: 1 | 2 | 3 | 4 | 5;
  comment: string;
};
export type ReviewsDataType = ReviewDataType[];
