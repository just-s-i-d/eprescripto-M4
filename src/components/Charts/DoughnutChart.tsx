import { useEffect, useState } from "react";

import { Card, Skeleton } from "antd";
import {
  ArcElement,
  Chart as ChartJs,
  Legend,
  Plugin,
  Tooltip,
} from "chart.js";
import { Doughnut } from "react-chartjs-2";
import { Chart, ChartOptions, plugins } from "chart.js";

import CardTitle from "@components/ui/CardTitle";
import ErrorBoundary from "@components/ErrorBoundary";
import { ApiResponseDataType, ChartDataType } from "@constants/types";
import useStatesHook from "src/hooks/useStatesHook";

ChartJs.register(ArcElement, Tooltip, Legend, plugins);
export const options: ChartOptions<"doughnut"> = {
  plugins: {
    legend: {
      display: true,
      position: "bottom",
    },
  },
};

type PieChartPropsType = {
  chartData?: ApiResponseDataType<ChartDataType>;
  title: string;
  label: string;
  error?: boolean;
  doughnutChart: ReturnType<
    typeof useStatesHook<ApiResponseDataType<ChartDataType>>
  >;
  darkMode?: boolean;
};
const colors: string[] = [
  "#96CCA3",
  "#3db5e8",
  "#3159a4",
  "#f7a825",
  "#ec6424",
];
const DoughnutChart = ({
  chartData,
  title,
  label,
  doughnutChart,
  error,
}: PieChartPropsType) => {
  const [loading, setLoading] = useState<boolean>(false);
  useEffect(() => {
    if (chartData) setLoading(false);
  }, [chartData]);
  const data = {
    labels: chartData?.attributes?.labels,
    datasets: [
      {
        label: label,
        data: chartData?.attributes?.data,
        backgroundColor: colors,
        borderColor: colors,
        borderWidth: 1,
      },
    ],
  };
  const plugins: Plugin<"doughnut">[] = [
    {
      id: "Doughnut Chart",
      beforeDraw: function (chart: Chart<"doughnut">) {
        const width = chart.width,
          height = chart.height,
          ctx = chart.ctx;
        ctx.restore();
        let fontSize = (height / 300).toFixed(2);
        ctx.font = fontSize + "em sans-serif";
        ctx.textBaseline = "top";
        const text1 = "Total Ratings",
          textX1 = Math.round((width - ctx.measureText(text1).width) / 2),
          textY1 = height / 2.9;
        ctx.fillText(text1, textX1, textY1);
        fontSize = (height / 100).toFixed(2);
        ctx.font = fontSize + "em sans-serif";
        const text2 = `${chartData?.attributes?.totalCount}`,
          textX2 = Math.round((width - ctx.measureText(text2).width) / 2),
          textY2 = height / 2.4;
        ctx.fillText(text2, textX2, textY2);
        ctx.save();
      },
    },
  ];
  return (
    <Card
      className="w-[32%] max-h-[60vh] max-xxl:w-[48%] max-xxl:h-[59vh] max-xl:w-[48%] max-md:w-9/12 max-sm:w-full"
      bordered={false}
    >
      <ErrorBoundary
        refreshComponent={() => doughnutChart.setRefresh((prev) => !prev)}
        error={error || doughnutChart.error}
      >
        <CardTitle>{title}</CardTitle>
        {loading ? (
          <Skeleton.Avatar active size={250} />
        ) : (
          <div className="h-[400px] flex justify-center items-center">
            <Doughnut
              data={data}
              options={options}
              plugins={plugins}
              data-testid="doughnut-chart"
            />
          </div>
        )}
      </ErrorBoundary>
    </Card>
  );
};
export default DoughnutChart;
