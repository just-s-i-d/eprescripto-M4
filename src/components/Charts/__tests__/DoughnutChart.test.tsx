import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import DoughnutChart from "../DoughnutChart";
import useStatesHook from "@hooks/useStatesHook";
import { ApiResponseDataType, ChartDataType } from "@constants/types";

jest.mock("@hooks/useStatesHook");

const mockedChartData = {
  id: 1,
  attributes: {
    labels: ["Label1", "Label2", "Label3", "Label4", "Label5"],
    data: [10, 20, 30, 15, 5],
    totalCount: 80,
  },
};
describe("DoughnutChart Component", () => {
  it("Test to check if it renders correctly", () => {
    (useStatesHook as jest.Mock).mockReturnValue({
      data: mockedChartData,
      error: false,
      loading: true,
      setRefresh: jest.fn(),
    });
    const doughnutChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
    const { asFragment } = render(
      <DoughnutChart
        darkMode={false}
        chartData={mockedChartData}
        title="Test Title"
        label="Test Label"
        doughnutChart={doughnutChart}
      />,
    );
    const chartElement = screen.getByTestId("doughnut-chart");
    expect(chartElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("handle error test", () => {
    (useStatesHook as jest.Mock).mockReturnValue({
      data: mockedChartData,
      error: true,
      loading: false,
      setRefresh: jest.fn(),
    });
    const doughnutChart = useStatesHook<ApiResponseDataType<ChartDataType>>();
    render(
      <DoughnutChart
        darkMode={false}
        error={true}
        chartData={mockedChartData}
        title="Test Title"
        label="Test Label"
        doughnutChart={doughnutChart}
      />,
    );
    const tryAgainBtn = screen.getByText("Try again");
    expect(screen.getByText("Something went wrong")).toBeInTheDocument();
    expect(tryAgainBtn).toBeInTheDocument();
    fireEvent.click(tryAgainBtn);
    expect(doughnutChart.setRefresh).toHaveBeenCalled();
  });
});
