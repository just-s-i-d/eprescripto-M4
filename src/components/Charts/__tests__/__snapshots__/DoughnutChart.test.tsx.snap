// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoughnutChart Component Test to check if it renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ant-card w-[32%] max-h-[60vh] max-xxl:w-[48%] max-xxl:h-[59vh] max-xl:w-[48%] max-md:w-9/12 max-sm:w-full css-dev-only-do-not-override-1adbn6x"
  >
    <div
      class="ant-card-body"
    >
      <h1
        class="text-xl font-bold mb-6 undefined"
      >
        Test Title
      </h1>
      <div
        class="h-[400px] flex justify-center items-center"
      >
        <canvas
          data-testid="doughnut-chart"
          height="150"
          role="img"
          width="300"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
